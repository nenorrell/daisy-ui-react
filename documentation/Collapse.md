⬅️ [Back to component index](README.md)

# Collapse

A collapsible container used for showing and hiding content.

### Props

| Prop               | Type                    | Default     | Description                                                                 |
| ------------------ | ----------------------- | ----------- | --------------------------------------------------------------------------- |
| `className`        | `string`                | `""`        | Custom classname to apply to the collapse container.                        |
| `defaultExpand`    | `boolean`               | `false`     | Determines whether the collapse should be expanded by default.              |
| `headerContent`    | `string` or `ReactNode` |             | The content for the collapse header.                                        |
| `headerClasses`    | `string`                | `""`        | Additional classname to apply to the collapse header.                       |
| `leftIcon`         | `ComponentOrJSX`        | `undefined` | The left icon for the collapse header.                                      |
| `leftIconClasses`  | `string`                | `""`        | Additional classname to apply to the left icon.                             |
| `rightIcon`        | `ComponentOrJSX`        | `undefined` | The right icon for the collapse header.                                     |
| `rightIconClasses` | `string`                | `""`        | Additional classname to apply to the right icon.                            |
| `parentCollapse`     | `boolean`               | `false`     | This is for when you're using nested collapses. Unfortunately, the expand animation does not currently work great for nested collapses. Marking this as true will mark this collapse as a parent & will disable the smooth expand for it (the parent) but will still allow for smooth expand for the children collapses.          |
| `rounded`          | `boolean`               | `false`     | Determines whether the collapse should have rounded corners or not.         |
| `onClick`          | `Function`              | `undefined` | Callback function that will be invoked when the collapse header is clicked. |


### Notes

* The `onClick` is only available on the headerContent and it will be called when the header is clicked
* Nested collapses are a little finnicky as things currently stand. Please see `parentCollapse` above for more details


### Examples

```javascript
import { Collapse } from 'daisy-ui-react-components';

function MyComponent() {
  const handleClick = e => {
    console.log('Collapse header clicked', e);
  }

  return (
    <div>
      <Collapse
        className="w-1/2"
        headerContent="Hello"
        headerClasses="bg-base-200 rounded-t-md"
        leftIcon={<FontAwesomeIcon icon={faGitlab} />}
        rightIcon={ArrowDownIcon}
      >
        <div className="bg-base-300">
          <p className="p-4">Hey everyone</p>
        </div>
      </Collapse>

      <Collapse headerContent="Click me to expand/collapse" onClick={handleClick}>
        <p>This is some expandable content</p>
      </Collapse>

      <Collapse headerContent="Click me to expand/collapse" >
        <p>This is some expandable content</p>
      </Collapse>

      <Collapse headerContent="Click me to expand/collapse" defaultExpand={true}>
        <p>This is some expandable content that is expanded by default</p>
      </Collapse>

      <Collapse headerContent={<div>Custom header content</div>} leftIcon={<MyCustomIcon />} className="my-custom-class">
        <p>This is some expandable content with custom header content and a custom left icon</p>
      </Collapse>
    </div>
  )
}
```