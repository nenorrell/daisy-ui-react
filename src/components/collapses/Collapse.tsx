import { ChevronDownIcon } from "@heroicons/react/24/solid";
import { RefObject, useEffect, useRef, useState } from "react";

const toggleExpand = (ref :RefObject<HTMLDivElement>, isExpanded :boolean, rootCollapse ?:boolean)=>{
    if(ref.current?.clientHeight && !isExpanded) {
        ref.current.style.height = "0px";
    }
    else if(ref.current && isExpanded) {
        if(!rootCollapse) {
            ref.current.style.height = "0px";
            ref.current.style.height = `${ref.current.scrollHeight}px`;
        }
    }
};

interface ICollapse {
    className ?:string;
    defaultExpand ?:boolean
    headerContent :string | React.ReactNode
    headerClasses ?:string
    LeftIcon ?:React.FunctionComponent<any>
    leftIconClasses ?:string
    RightIcon ?:React.FunctionComponent<any>
    rightIconClasses ?:string
    rootCollapse ?:boolean
    rounded ?:boolean
    onClick ?:Function
}
export const Collapse = (props :React.PropsWithChildren<ICollapse>) => {
    const expandableContent = useRef<HTMLDivElement>(null);
    const [isExpanded, setIsExpanded] = useState<boolean>(props.defaultExpand || false);
    useEffect(()=>{
        toggleExpand(expandableContent, isExpanded, props.rootCollapse);
    }, [isExpanded]); // eslint-disable-line

    return (
        <div tabIndex={0} className={`w-100 collapse ${props.className || ""}`}>
            <div className={`collapse-title cursor-pointer flex items-center justify-center px-4 ${props.headerClasses || ""}`}
                onClick={(e)=>{
                    setIsExpanded(!isExpanded);
                    props.onClick && props.onClick(e);
                } }>
                {
                    props.LeftIcon && <props.LeftIcon className={props.leftIconClasses || "w-6 h-6 mr-3"} />
                }
                <div className="w-full">{props.headerContent}</div>
                <div className={`items-end transition ease-in-out ${isExpanded ? "" : "-rotate-90"}`}>
                    {
                        props.RightIcon &&
                        <props.RightIcon className={props.rightIconClasses || "w-4 h-4"} /> ||
                        <ChevronDownIcon className={props.rightIconClasses || "w-4 h-4"}/>
                    }
                </div>
            </div>

            <div
                ref={expandableContent}
                className={`collapse-body ease-linear duration-[.25s] transition-[height] overflow-hidden ${isExpanded ? "" : "h-0"}`}
            >
                {props.children}
            </div>
        </div>
    );
};
