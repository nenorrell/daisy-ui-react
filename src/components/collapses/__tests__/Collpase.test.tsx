import { HomeIcon } from "@heroicons/react/24/solid";
import { render, fireEvent } from "@testing-library/react";
import { Collapse } from "../Collapse";

describe("Collapse Component", () => {
    beforeEach(() => {
        Object.defineProperty(HTMLElement.prototype, "scrollHeight", { configurable: true, value: 30 });
        Object.defineProperty(HTMLElement.prototype, "clientHeight", { configurable: true, value: 10 });
    });

    afterEach(()=>{
        // Set these back to undefined to prevent test bleed
        delete (HTMLElement as any).prototype.scrollHeight;
        delete (HTMLElement as any).prototype.clientHeight;
    });

    it("Renders a collapse component", () => {
        const component = render(
            <Collapse
                headerContent='My Collapse'
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const collapseBody = component.container.querySelector<HTMLDivElement>("div.collapse-body");
        expect(collapseBody?.classList.contains("h-0")).toBe(true);
        expect(collapseBody?.style.height).toBe("0px");
        expect(component.container).toMatchSnapshot();
    });

    it("Respects onClick when passed in", ()=>{
        const stub = jest.fn();
        const component = render(
            <Collapse
                headerContent='My Collapse'
                onClick={stub}
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const collapseTitle = component.container.querySelector<HTMLDivElement>("div.collapse-title");
        fireEvent.click(collapseTitle as HTMLDivElement);
        expect(stub).toHaveBeenCalled();
    });

    it("Toggles collapse on click", () => {
        const component = render(
            <Collapse
                headerContent='My Collapse'
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const collapseTitle = component.container.querySelector<HTMLDivElement>("div.collapse-title");
        const collapseBody = component.container.querySelector<HTMLDivElement>("div.collapse-body");
        const chevronWrapper = collapseTitle?.querySelector<HTMLDivElement>("div.items-end");

        // Initial state
        expect(collapseBody?.classList.contains("h-0")).toBe(true);
        expect(collapseBody?.style.height).toBe("0px");
        expect(chevronWrapper?.classList.contains("-rotate-90")).toBe(true);

        // Expand
        fireEvent.click(collapseTitle as HTMLDivElement);
        expect(collapseBody?.classList.contains("h-0")).toBe(false);
        expect(collapseBody?.style.height).toBe("30px");
        expect(chevronWrapper?.classList.contains("-rotate-90")).toBe(false);

        // Collapse
        fireEvent.click(collapseTitle as HTMLDivElement);
        expect(collapseBody?.classList.contains("h-0")).toBe(true);
        expect(collapseBody?.style.height).toBe("0px");
        expect(chevronWrapper?.classList.contains("-rotate-90")).toBe(true);
    });

    it("Toggles collapse on click when defaultExpand is set to true", () => {
        const component = render(
            <Collapse
                defaultExpand={true}
                headerContent='My Collapse'
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );
        const collapseTitle = component.container.querySelector<HTMLDivElement>("div.collapse-title");
        const collapseBody = component.container.querySelector<HTMLDivElement>("div.collapse-body");
        const chevronWrapper = collapseTitle?.querySelector<HTMLDivElement>("div.items-end");

        // Initial state
        expect(collapseBody?.classList.contains("h-0")).toBe(false);
        expect(collapseBody?.style.height).toBe("30px");
        expect(chevronWrapper?.classList.contains("-rotate-90")).toBe(false);

        // Collapse
        fireEvent.click(collapseTitle as HTMLDivElement);
        expect(collapseBody?.classList.contains("h-0")).toBe(true);
        expect(collapseBody?.style.height).toBe("0px");
        expect(chevronWrapper?.classList.contains("-rotate-90")).toBe(true);

        // Expand
        fireEvent.click(collapseTitle as HTMLDivElement);
        expect(collapseBody?.classList.contains("h-0")).toBe(false);
        expect(collapseBody?.style.height).toBe("30px");
        expect(chevronWrapper?.classList.contains("-rotate-90")).toBe(false);
    });

    it("Renders an expanded collapse component when defaultExpand is true", () => {
        const component = render(
            <Collapse
                headerContent='My Collapse'
                defaultExpand={true}
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );
        const collapseBody = component.container.querySelector<HTMLDivElement>("div.collapse-body");

        expect(collapseBody?.classList.contains("h-0")).toBe(false);
        expect(collapseBody?.style.height).toBe("30px");
        expect(component.container).toMatchSnapshot();
    });

    it("Respects className prop", () => {
        const {container} = render(
            <Collapse
                headerContent='My Collapse'
                className='my-class-name'
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const el = container.firstChild as HTMLElement;
        expect(el).toBeInTheDocument();
        expect(el.classList.contains("my-class-name")).toBe(true);
        expect(el).toMatchSnapshot();
    });

    it("Respects headerClasses prop", () => {
        const {container} = render(
            <Collapse
                headerContent='My Collapse'
                headerClasses='some-header-class another-header-class'
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );
        const headerContainer = container.querySelector("div.some-header-class.another-header-class");
        expect(headerContainer).toBeInTheDocument();
        expect(headerContainer).toMatchSnapshot();
    });

    it("Respects leftIcon prop as a functional component", () => {
        const {container} = render(
            <Collapse
                headerContent='My Collapse'
                leftIcon={HomeIcon}
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const collapseTitleEl = container.querySelector("div.collapse-title");
        expect(collapseTitleEl?.querySelector("svg")).toEqual(collapseTitleEl?.firstElementChild);
        expect(collapseTitleEl).toMatchSnapshot();
    });

    it("Respects leftIcon prop as a JSX element", () => {
        const {container} = render(
            <Collapse
                headerContent='My Collapse'
                leftIcon={<HomeIcon />}
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const collapseTitleEl = container.querySelector("div.collapse-title");
        expect(collapseTitleEl?.querySelector("svg")).toEqual(collapseTitleEl?.firstElementChild);
        expect(collapseTitleEl).toMatchSnapshot();
    });

    it("Respects rightIcon prop as a functional component", () => {
        const {container} = render(
            <Collapse
                headerContent='My Collapse'
                rightIcon={HomeIcon}
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const rightIconContainer = container.querySelector("div.items-end");
        expect(rightIconContainer?.querySelector("svg")).toEqual(rightIconContainer?.firstElementChild);
        expect(rightIconContainer).toMatchSnapshot();
    });

    it("Respects rightIcon prop as a JSX element", () => {
        const {container} = render(
            <Collapse
                headerContent='My Collapse'
                rightIcon={<HomeIcon />}
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const rightIconContainer = container.querySelector("div.items-end");
        expect(rightIconContainer?.querySelector("svg")).toEqual(rightIconContainer?.firstElementChild);
        expect(rightIconContainer).toMatchSnapshot();
    });

    it("Does not render leftIcon when prop is undefined", () => {
        const {container} = render(
            <Collapse
                headerContent='My Collapse'
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const collapseTitleEl = container.querySelector("div.collapse-title");
        expect(collapseTitleEl?.querySelector("svg")).not.toEqual(collapseTitleEl?.firstElementChild);
        expect(collapseTitleEl).toMatchSnapshot();
    });

    it("Respects leftIconClasses arg", () => {
        const {container} = render(
            <Collapse
                headerContent='My Collapse'
                leftIcon={HomeIcon}
                leftIconClasses='some-icon-class'
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const collapseTitleEl = container.querySelector("div.collapse-title");
        expect(collapseTitleEl?.querySelector("svg.some-icon-class")).toBeInTheDocument();
        expect(collapseTitleEl).toMatchSnapshot();
    });

    it("Respects rightIconClasses arg", () => {
        const {container} = render(
            <Collapse
                headerContent='My Collapse'
                rightIconClasses='some-icon-class'
            >
                <p className="test-element">Hi everyone</p>
            </Collapse>
        );

        const collapseTitleEl = container.querySelector("div.collapse-title");
        const iconEl = collapseTitleEl?.querySelector("svg.some-icon-class");
        expect(iconEl).toBeInTheDocument();
        expect(collapseTitleEl).toMatchSnapshot();
    });
});
