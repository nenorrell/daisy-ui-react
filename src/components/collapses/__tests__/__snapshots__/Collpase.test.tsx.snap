// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse Component Does not render LeftIcon when prop is undefined 1`] = `
<div
  class="collapse-title cursor-pointer flex items-center justify-center px-4 "
>
  <div
    class="w-full"
  >
    My Collapse
  </div>
  <div
    class="items-end transition ease-in-out -rotate-90"
  >
    <svg
      aria-hidden="true"
      class="w-4 h-4"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;

exports[`Collapse Component Renders a collapse component 1`] = `
<div>
  <div
    class="w-100 collapse "
    tabindex="0"
  >
    <div
      class="collapse-title cursor-pointer flex items-center justify-center px-4 "
    >
      <div
        class="w-full"
      >
        My Collapse
      </div>
      <div
        class="items-end transition ease-in-out -rotate-90"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      class="collapse-body ease-linear duration-[.25s] transition-[height] overflow-hidden h-0"
      style="height: 0px;"
    >
      <p
        class="test-element"
      >
        Hi everyone
      </p>
    </div>
  </div>
</div>
`;

exports[`Collapse Component Renders an expanded collapse component when defaultExpand is true 1`] = `
<div>
  <div
    class="w-100 collapse "
    tabindex="0"
  >
    <div
      class="collapse-title cursor-pointer flex items-center justify-center px-4 "
    >
      <div
        class="w-full"
      >
        My Collapse
      </div>
      <div
        class="items-end transition ease-in-out "
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      class="collapse-body ease-linear duration-[.25s] transition-[height] overflow-hidden "
      style="height: 30px;"
    >
      <p
        class="test-element"
      >
        Hi everyone
      </p>
    </div>
  </div>
</div>
`;

exports[`Collapse Component Respects className prop 1`] = `
<div
  class="w-100 collapse my-class-name"
  tabindex="0"
>
  <div
    class="collapse-title cursor-pointer flex items-center justify-center px-4 "
  >
    <div
      class="w-full"
    >
      My Collapse
    </div>
    <div
      class="items-end transition ease-in-out -rotate-90"
    >
      <svg
        aria-hidden="true"
        class="w-4 h-4"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
  <div
    class="collapse-body ease-linear duration-[.25s] transition-[height] overflow-hidden h-0"
    style="height: 0px;"
  >
    <p
      class="test-element"
    >
      Hi everyone
    </p>
  </div>
</div>
`;

exports[`Collapse Component Respects headerClasses prop 1`] = `
<div
  class="collapse-title cursor-pointer flex items-center justify-center px-4 some-header-class another-header-class"
>
  <div
    class="w-full"
  >
    My Collapse
  </div>
  <div
    class="items-end transition ease-in-out -rotate-90"
  >
    <svg
      aria-hidden="true"
      class="w-4 h-4"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;

exports[`Collapse Component Respects leftIcon prop 1`] = `
<div
  class="collapse-title cursor-pointer flex items-center justify-center px-4 "
>
  <svg
    aria-hidden="true"
    class="w-6 h-6 mr-3"
    fill="currentColor"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M11.47 3.84a.75.75 0 011.06 0l8.69 8.69a.75.75 0 101.06-1.06l-8.689-8.69a2.25 2.25 0 00-3.182 0l-8.69 8.69a.75.75 0 001.061 1.06l8.69-8.69z"
    />
    <path
      d="M12 5.432l8.159 8.159c.03.03.06.058.091.086v6.198c0 1.035-.84 1.875-1.875 1.875H15a.75.75 0 01-.75-.75v-4.5a.75.75 0 00-.75-.75h-3a.75.75 0 00-.75.75V21a.75.75 0 01-.75.75H5.625a1.875 1.875 0 01-1.875-1.875v-6.198a2.29 2.29 0 00.091-.086L12 5.43z"
    />
  </svg>
  <div
    class="w-full"
  >
    My Collapse
  </div>
  <div
    class="items-end transition ease-in-out -rotate-90"
  >
    <svg
      aria-hidden="true"
      class="w-4 h-4"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;

exports[`Collapse Component Respects leftIconClasses arg 1`] = `
<div
  class="collapse-title cursor-pointer flex items-center justify-center px-4 "
>
  <svg
    aria-hidden="true"
    class="some-icon-class"
    fill="currentColor"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M11.47 3.84a.75.75 0 011.06 0l8.69 8.69a.75.75 0 101.06-1.06l-8.689-8.69a2.25 2.25 0 00-3.182 0l-8.69 8.69a.75.75 0 001.061 1.06l8.69-8.69z"
    />
    <path
      d="M12 5.432l8.159 8.159c.03.03.06.058.091.086v6.198c0 1.035-.84 1.875-1.875 1.875H15a.75.75 0 01-.75-.75v-4.5a.75.75 0 00-.75-.75h-3a.75.75 0 00-.75.75V21a.75.75 0 01-.75.75H5.625a1.875 1.875 0 01-1.875-1.875v-6.198a2.29 2.29 0 00.091-.086L12 5.43z"
    />
  </svg>
  <div
    class="w-full"
  >
    My Collapse
  </div>
  <div
    class="items-end transition ease-in-out -rotate-90"
  >
    <svg
      aria-hidden="true"
      class="w-4 h-4"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;

exports[`Collapse Component Respects rightIconClasses arg 1`] = `
<div
  class="collapse-title cursor-pointer flex items-center justify-center px-4 "
>
  <div
    class="w-full"
  >
    My Collapse
  </div>
  <div
    class="items-end transition ease-in-out -rotate-90"
  >
    <svg
      aria-hidden="true"
      class="some-icon-class"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;
